components/service/nimbus/src/main/java/mozilla/components/service/nimbus/Nimbus.kt

```diff
@@ -0,0 +1,211 @@
+ */
+internal interface NimbusApi {
+    /**
+     * Get the list of currently enrolled experiments
+     *
+     * @return A list of [EnrolledExperiment]s
+     */
+    fun getActiveExperiments(): List<EnrolledExperiment>
+
+    /**
+     * Get the currently enrolled branch for the given experiment
+     *
+     * @param experimentId The string experiment-id or "slug" for which to retrieve the branch
+     *
+     * @return A String representing the branch-id or "slug"
+     */
+    fun getExperimentBranch(experimentId: String): String?
+
+    /**
+     * Refreshes the experiments from the endpoint.
+     */
+    fun updateExperiments()
+
+    /**
+     * Opt out of a specific experiment
+     *
+     * @param experimentId The string experiment-id or "slug" for which to opt out of
+     */
+    fun optOut(experimentId: String)
+
+    /**
+     * Opt out of all experiments
+     */
+    fun optOutAll()
+}
+
+/**
+ * A singleton implementation of the [NimbusApi] interface backed by the Nimbus SDK.
+ */
+object Nimbus : NimbusApi {
+    private const val LOG_TAG = "service/Nimbus"
+    private const val EXPERIMENT_COLLECTION_NAME = "nimbus-mobile-experiments"
+    private const val NIMBUS_DATA_DIR: String = "nimbus_data"
+
+    // Using a single threaded executor here to enforce synchronization where needed.
+    private val scope: CoroutineScope =
+        CoroutineScope(Executors.newSingleThreadExecutor().asCoroutineDispatcher())
+
+    private lateinit var nimbus: NimbusClient
+    private lateinit var dataDir: File
+    private var onExperimentUpdated: ((List<EnrolledExperiment>) -> Unit)? = null
+
+    private var isInitialized = false
+
+    /**
+     * Initialize the Nimbus SDK library.
+     *
+     * This should only be initialized once by the application.
+     *
+     * @param context [Context] to access application and device parameters.  As we cannot enforce
+     * through the compiler that the context pass to the initialize function is a Application
+     * Context, there could potentially be a memory leak if the initializing application doesn't
+     * comply.
+     *
+     * @param onExperimentUpdated callback that will be executed when the list of experiments has
+     * been updated from the experiments endpoint and evaluated by the Nimbus-SDK. This is meant to
+     * be used for consuming applications to perform any actions as a result of enrollment in an
+     * experiment so that the application is not required to await the network request. The
+     * callback will be supplied with the list of active experiments (if any) for which the client
+     * is enrolled.
+     */
+    fun init(
+        context: Context,
+        onExperimentUpdated: ((activeExperiments: List<EnrolledExperiment>) -> Unit)? = null
+    ) {
+        // Set the name of the native library
+        System.setProperty("uniffi.component.nimbus.libraryOverride", "megazord")
```

Comment:
Oh, actually, this won't work correctly with custom megazords which may use a different name for the dynamic library file. We might need to figure out how to put something like https://github.com/mozilla/glean/pull/1290 into uniffi and/or nimbus so that we can hook up this logic at build time.

Reply:
(I don't think we have any immediate consumers of nimbus who would be using a custom megazord, but if we get one in future they'll be looking at some pretty surprising runtime symbol lookup failures)

Reply:
I filed https://github.com/mozilla/uniffi-rs/issues/343 to track improving this on the uniffi side of things.

For now, what do you think about allowing the application to override this value? I'm imagining something like:

```
if (System.getProperty("uniffi.component.nimbus.libraryOverride") == null) {
    System.setProperty("uniffi.component.nimbus.libraryOverride", "megazord")
}
```

And then an application that is using a custom megazord could set its own value for this property if needed.